File: api-khairat/src/main/java/com/izan/khairatapi/service/DependentService.java
package com.izan.khairatapi.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.izan.khairatapi.entity.Dependent;
import com.izan.khairatapi.entity.Member;
import com.izan.khairatapi.entity.Person;
import com.izan.khairatapi.repository.DependentRepository;
import com.izan.khairatapi.repository.MemberRepository;
import com.izan.khairatapi.repository.PersonRepository;

@Service
public class DependentService {
    
    @Autowired
    private DependentRepository dependentRepository;
    @Autowired
    private PersonRepository personRepository;
    @Autowired
    private MemberRepository memberRepository;

    public Dependent save(Dependent dependent, Long memberId) {
        Person newPerson = dependent.getPerson();
        newPerson = personRepository.save(newPerson);
        dependent.setPerson(newPerson);
        Member member = memberRepository.findById(memberId).orElse(null);
        dependent.setMember(member);
        return dependentRepository.save(dependent);
    }

    public void deleteById(Long id) {
        Dependent dependent = dependentRepository.findById(id).orElse(null);
        Person person = personRepository.findById(dependent.getPerson().getId()).orElse(null);
        dependentRepository.delete(dependent);
        personRepository.delete(person);
    }
}

File: legacy/e-masjid.my-2.0/api/khairat-api/service/dependent_service.go
package service

import (
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/model"
	"github.com/Dev4w4n/e-masjid.my/api/khairat-api/repository"
	"github.com/gin-gonic/gin"
)

type DependentService interface {
	Save(ctx *gin.Context, dependent model.Dependent, memberId int) error
	Delete(ctx *gin.Context, d int) error
}

type DependentServiceImpl struct {
	dependentRepository repository.DependentRepository
	memberRepository    repository.MemberRepository
	personRepository    repository.PersonRepository
}

func NewDependentService(dependentRepository repository.DependentRepository, memberRepository repository.MemberRepository, personRepository repository.PersonRepository) DependentService {
	return &DependentServiceImpl{
		dependentRepository: dependentRepository,
		memberRepository:    memberRepository,
		personRepository:    personRepository,
	}
}

// Delete implements DependentService.
func (repo *DependentServiceImpl) Delete(ctx *gin.Context, id int) error {
	dependent, err := repo.dependentRepository.FindById(ctx, id)

	if err != nil {
		return err
	}

	person, err := repo.personRepository.FindById(ctx, dependent.Person.Id)

	if err != nil {
		return err
	}

	err = repo.dependentRepository.Delete(ctx, dependent)

	if err != nil {
		return err
	}

	err = repo.personRepository.Delete(ctx, person.Id)

	if err != nil {
		return err
	}

	return nil
}

// Save implements DependentService.
func (repo *DependentServiceImpl) Save(ctx *gin.Context, dependent model.Dependent, memberId int) error {
	newPerson, err := repo.personRepository.Save(ctx, dependent.Person)
	if err != nil {
		return err
	}
	dependent.Person = newPerson

	member, err := repo.memberRepository.FindById(ctx, memberId)
	if err != nil {
		return err
	}
	dependent.Member = member

	err = repo.dependentRepository.Save(ctx, dependent, int64(memberId))
	if err != nil {
		return err
	}

	return nil
}
